### PR TITLE
Fix the None Object Issue Permanently,  #1243 #1240 #1233 #1228 #1225 #1218

### PR DESCRIPTION
**Pull request by ifahadone**
resolved these #1243 #1240 #1233 #1228 #1225 #1218 , as of 2022-02-08

### CODE REVIEW

**Review by Bryan-Roger — APPROVED** [review]

**Review by vishnudas-bluefox — APPROVED** [review]

**Comment by Panos-Jr** [comment]
Excellent!

File: pytube/cipher.py [hunk 1]
```diff
@@ -273,7 +273,16 @@ def get_throttling_function_name(js: str) -> str:
         function_match = regex.search(js)
         if function_match:
             logger.debug("finished regex search, matched: %s", pattern)
-            return function_match.group(1)
+            function_name = function_match.group(1)
+            is_Array = True if '[' or ']' in function_name else False
```
Comment (glubsy) [inline pytube/cipher.py]: This is wrong, as pointed out by [this comment](https://github.com/pytube/pytube/issues/1243#issuecomment-1037010471).

**Review by glubsy — CHANGES_REQUESTED** [review]
You need to add proper unit tests.

File: pytube/cipher.py (hunk 1 above)
Comment (glubsy) [reply pytube/cipher.py]: Also why make a variable out of it?

**Review by ifahadone — COMMENTED** [review]
fixed the is_array var

File: pytube/cipher.py [hunk 2]
```diff
@@ -273,7 +273,16 @@ def get_throttling_function_name(js: str) -> str:
...
+            function_name = function_match.group(1)
+            is_Array = True if '[' in function_name or ']' in function_name else False
+            if is_Array:
+                index = int(re.findall(r'\d+', function_name)[0])
+                name = function_name.split('[')[0]
+                pattern = r"var %s=\[(.*?)\];" % name
+                regex = re.compile(pattern)
+                return regex.search(js).group(1).split(',')[index]
+            else:
+                return function_name
```
Comment (SlavaSkvortsov) [inline pytube/cipher.py]:
```suggestion
            is_array = '[' in function_name or ']' in function_name
            if not is_array:
                return function_name
                
            index = int(re.findall(r'\d+', function_name)[0])
            name = function_name.split('[')[0]
            regex = re.compile(r"var %s=\[(.*?)\];" % name)
            return regex.search(js).group(1).split(',')[index]
```

**Comment by LonelyGuy12** [comment]
Amazing! please merge it soon!

**Review by berke-horozal-crea — APPROVED** [review]

**Comment by huynguyencong** [comment]
Excellent! Please merge this one.

**Comment by RONNCC** [comment]
hey @ifahadone  makes sense -- can you add a unit test for it / fix the unit tests and we can merge

**Comment by RONNCC** [comment]
@glubsy  can you look at it -- also can you add a test @ifahadone  :-) -- sorry the youtube code change ups are definitely not the easiest to deal with

**Review by glubsy — COMMENTED** [review]
Looks functional, but it is essentially the same code as #1222, albeit a little simpler to read.

**Comment by glubsy** [comment]
I believe this can be closed. Still, thanks for the submission @ifahadone.

**Comment by RONNCC** [comment]
@glubsy  is the issue already fixed in main?

**Comment by glubsy** [comment]
> @glubsy is the issue already fixed in main?

I believe so.

**Comment by Martouta** [comment]
@RONNCC @glubsy Not really. It happened again and now there are many PRs open for this issue. But this PR is supposed to address it permanently (according to the title of the PR) 👍🏼

**Comment by glubsy** [comment]
@Martouta it won't fix the issue permanently. 

The follow up to this issue is fixed by #1282. All the other PRs are spam and should be deleted.

**Comment by RONNCC** [comment]
@Martouta  to elaborate a bit the issue is caused by Youtube changing their code - hence it can break at any time since by definition it's not a stable interface :-)

**Pull request closed without merging**